keanu-project/src/main/java/io/improbable/keanu/vertices/dbltensor/probabilistic/ProbabilisticVertexShaping.java

```diff
@@ -0,0 +1,61 @@
+package io.improbable.keanu.vertices.dbltensor.probabilistic;
+
+import io.improbable.keanu.vertices.dbltensor.DoubleTensor;
+import io.improbable.keanu.vertices.dbltensor.TensorShape;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+public class ProbabilisticVertexShaping {
+
+    /**
+     * This is a common function to check that a vertex's parent tensors are either
+     * the same shape of the vertex in question OR scalar.
+     *
+     * @param shape   the shape of the vertex being checked
+     * @param tensors the tensors of the parent vertices
+     */
+    public static void checkParentShapes(int[] shape, DoubleTensor... tensors) {
+
+        Set<TensorShape> nonScalarShapes = getNonScalarShapes(tensors);
+
+        if (!nonScalarShapes.isEmpty()) {
+
+            boolean moreThanOneNonScalarShape = nonScalarShapes.size() > 1;
+            boolean nonScalarShapeDoesNotMatchProposal = !Arrays.equals(nonScalarShapes.iterator().next().getShape(), shape);
+
+            if (moreThanOneNonScalarShape || nonScalarShapeDoesNotMatchProposal) {
+                throw new IllegalArgumentException("invalid hyper parameter shapes");
+            }
+        }
+    }
+
+    /**
+     * This implies a vertex's tensor shape based on it's parents tensor shapes.
+     *
+     * @param tensors the parent tensors of the vertex that needs it's shape implied.
+     * @return a suggested shape for the vertex given it's parent vertices shapes.
```

Comment:
I think we should use stronger language here since we throw an exception if it's wrong. Something like `the shape the vertex must be given its parent vertices shapes`

Reply:
Good call. I've added some stronger wording and the fact that it can throw an exception.